File: leetcodejvm/src/main/java/linkedlist/ReverseLinkedList.java
package linkedlist;

import base.Node;

public class ReverseLinkedList {
    public Node reverseList(Node head) {
        if (head == null || head.next == null) {
            return head;
        }

        Node dummy = new Node(0);

        dummy.next = head;
        Node cur = head;
        Node next = head.next;

        while (next != null) {
            cur.next = next.next;
            next.next = dummy.next;
            dummy.next = next;
            next = cur.next;
        }
        return dummy.next;
    }
}

File: leetcodejvm/src/main/kotlin/linkedlist/ReverseLinkedList.kt
package linkedlist

import datastructure.ListNode

class ReverseLinkedList {
    fun reverse(head: ListNode?): ListNode? {
        if (head?.next == null) {
            return head
        }
        var prev: ListNode? = null
        var cur: ListNode? = head
        var next: ListNode? = null
        while (cur != null) {
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        }
        return prev
    }
}